samples/canopen/canopen_sdo/main.cpp

```diff
@@ -0,0 +1,122 @@
+                    message.getId(),
+                    message.getDataLength(),
+                    messageString);
+}
+
+int main() {
+    // Initialize system
+    core::platform::init();
+
+    io::UART& uart = io::getUART<io::Pin::UART_TX, io::Pin::UART_RX>(9600);
+    log::LOGGER.setUART(&uart);
+    log::LOGGER.setLogLevel(log::Logger::LogLevel::DEBUG);
+
+    // create the SDO node
+    SDOCanNode testCanNode;
+
+    dev::Timer& timer = dev::getTimer<dev::MCUTimer::Timer2>(100);
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Setup CAN configuration, this handles making drivers, applying settings.
+    // And generally creating the CANopen stack node which is the interface
+    // between the application (the code we write) and the physical CAN network
+    ///////////////////////////////////////////////////////////////////////////
+
+    // Will store CANopen messages that will be populated by the EVT-core CAN
+    // interrupt
+    core::types::FixedQueue<CANOPEN_QUEUE_SIZE, io::CANMessage> canOpenQueue;
+
+    // Initialize CAN, add an IRQ which will add messages to the queue above
+    io::CAN& can = io::getCAN<io::Pin::PA_12, io::Pin::PA_11>();
+    can.addIRQHandler(canInterrupt, reinterpret_cast<void*>(&canOpenQueue));
+
+    // Reserved memory for CANopen stack usage
+    uint8_t sdoBuffer[CO_SSDO_N * CO_SDO_BUF_BYTE];
+    CO_TMR_MEM appTmrMem[16];
+
+    // Reserve driver variables
+    CO_IF_DRV canStackDriver;
+
+    CO_IF_CAN_DRV canDriver;
+    CO_IF_TIMER_DRV timerDriver;
+    CO_IF_NVM_DRV nvmDriver;
+
+    CO_NODE canNode;
+
+    // Attempt to join the CAN network
+    io::CAN::CANStatus result = can.connect();
+
+    // test that the board is connected to the can network
+    if (result != io::CAN::CANStatus::OK) {
+        uart.printf("Failed to connect to CAN network\r\n");
+        return 1;
+    }
+
+    // Initialize all the CANOpen drivers.
+    io::initializeCANopenDriver(&canOpenQueue, &can, &timer, &canStackDriver, &nvmDriver, &timerDriver, &canDriver);
+
+    // Initialize the CANOpen node we are using.
+    io::initializeCANopenNode(&canNode, &testCanNode, &canStackDriver, sdoBuffer, appTmrMem);
+
+    // Set the node to operational mode
+    CONmtSetMode(&canNode.Nmt, CO_OPERATIONAL);
+
+    time::wait(500);
+
+    // print any CANopen errors
+    uart.printf("Error: %d\r\n", CONodeGetErr(&canNode));
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Main loop
+    ///////////////////////////////////////////////////////////////////////////
+    uint32_t lastUpdate1 = HAL_GetTick();
+    uint32_t lastUpdate2 = HAL_GetTick();
+
+    while (1) {
+        if ((HAL_GetTick() - lastUpdate1) >= 1000) {        // If 1000ms have passed receive CAN message.
+            testCanNode.receiveData(canNode);               // Receive data from server
```

Comment:
I really do not like that you have to pass the CO_NODE in every time you call this function. I think it might be better to store the CO_NODE in SDOCanNode, but I'm not sure because you'd have to make a function to get a pointer to pass into `io::processCANopenNode`. @mjh9585 what do you think?

Reply:
I think it is fine as you still have to pass it into `io::processCANopenNode` and an extra method to access CO_NODE from SDOCanNode would make it harder to follow. Perhaps a better approach would be to pass a pointer to CO_NODE into the constructor instead.